ArduPlane/fence.cpp

```diff
@@ -0,0 +1,132 @@
+
+    // we still don't do anything when disarmed, but we do check for fence breaches.
+    // fence pre-arm check actually checks if any fence has been breached
+    // that's not ever going to be true if we don't call check on AP_Fence while disarmed
+    if (!arming.is_armed())
+    {
+        return;
+    }
+
+    if (!fence.enabled())
+    {
+        // Switch back to the chosen control mode if still in
+        // GUIDED to the return point
+        if ((fence.get_action() == FENCE_ACTION_GUIDED || fence.get_action() == FENCE_ACTION_GUIDED_THR_PASS || fence.get_action() == FENCE_ACTION_RTL) &&
+            (control_mode == &mode_guided || control_mode == &mode_avoidADSB))
+        {
+            set_mode(*previous_mode, ModeReason::GCS_COMMAND);
+        }
+    }
+
+    // Never trigger a fence breach in the final stage of landing
+    if (landing.is_expecting_impact())
+    {
+        return;
+    }
+
+    if (new_breaches)
+    {
+        // if the user wants some kind of response and motors are armed
+        const uint8_t fence_act = fence.get_action();
+        switch (fence_act)
+        {
+        case AC_FENCE_ACTION_REPORT_ONLY:
+            break;
+        case AC_FENCE_ACTION_GUIDED:
+        case AC_FENCE_ACTION_GUIDED_THROTTLE_PASS:
+        case AC_FENCE_ACTION_RTL_AND_LAND:
+            // make sure we don't auto trim the surfaces on this mode change
+            int8_t saved_auto_trim = g.auto_trim;
+            g.auto_trim.set(0);
+            if (fence_act == AC_FENCE_ACTION_RTL_AND_LAND) {
+                set_mode(mode_rtl, ModeReason::FENCE_BREACHED);
+            } else {
+                set_mode(mode_guided, ModeReason::FENCE_BREACHED);
+            }
+            g.auto_trim.set(saved_auto_trim);
+
+            if (fence.get_return_rally() != 0 || fence_act == AC_FENCE_ACTION_RTL_AND_LAND) {
+                guided_WP_loc = rally.calc_best_rally_or_home_location(current_loc, get_RTL_altitude());
+            } else {
+                //return to fence return point, not a rally point
+                guided_WP_loc = {};
+                if (fence.get_return_altitude() > 0) {
+                    // fly to the return point using _retalt
+                    guided_WP_loc.alt = home.alt + 100.0f * fence.get_return_altitude();
+                } else if (fence.get_safe_alt_min() >= fence.get_safe_alt_max()) {
+                    // invalid min/max, use RTL_altitude
+                    guided_WP_loc.alt = home.alt + g.RTL_altitude_cm;
+                } else {
+                    // fly to the return point, with an altitude half way between
+                    // min and max
+                    guided_WP_loc.alt = home.alt + 100.0f * (fence.get_safe_alt_min() + fence.get_safe_alt_max()) / 2;
+                }
+
+                AC_PolyFenceItem return_point;
+                //! TODO: Determine if this works?
+                if (fence.polyfence().get_item(fence.polyfence().num_stored_items() - 1, return_point) &&
```

Comment:
Check my previous PR where I synthesises a return point if we didn't have one.... basically the geometric centre (read "average location") of all of the fencepoints.  Which, yes, can be outside the fence....

Reply:
Awesome, I've nabbed the `polyfence().get_return_point(...)` to find the geometric centre. Thanks. 😄